### PR TITLE
feat: migrate file engine from table to reigon

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Introduce FileRegionEngine. Remove FileTableEngine.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/file-engine/src/manifest.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            .context(DeleteRegionManifestSnafu { region_id })
+    }
+
+    pub fn metadata(&self) -> Result<RegionMetadataRef> {
+        let mut builder = RegionMetadataBuilder::new(self.region_id);
+        for column in &self.column_metadatas {
+            builder.push_column_metadata(column.clone());
+        }
+        builder.primary_key(self.primary_key.clone());
+        let metadata = builder.build().context(InvalidMetadataSnafu)?;
```
Comment (WenyXu) [inline src/file-engine/src/manifest.rs:96]: The builder will fail if we don't contain any columns that have the `Timestamp` semantic.

File: src/file-engine/src/engine.rs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+        let res = FileRegion::create(region_id, request, &self.object_store).await;
+        let region = res.inspect_err(|err| {
+            error!(
+                "Failed to create region, region_id: {}, err: {}",
+                region_id, err
+            );
+        })?;
+        self.regions.write().await.insert(region_id, region);
+
+        info!("A new region created, region_id: {}", region_id);
```
Comment (waynexia) [inline src/file-engine/src/engine.rs:188]: nit:
```suggestion
        info!("A new region is created, region_id: {}", region_id);
```

File: src/file-engine/src/manifest.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+#[inline]
+fn region_manifest_path(region_dir: &str) -> String {
+    format!("{region_dir}manifest/_immutable_manifest")
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct FileRegionManifest {
+    pub region_id: RegionId,
+    pub column_metadatas: Vec<ColumnMetadata>,
+    pub primary_key: Vec<ColumnId>,
```
Comment (waynexia) [inline src/file-engine/src/manifest.rs:41]: Does file engine need primary keys? cc @WenyXu

Comment (WenyXu) [reply src/file-engine/src/manifest.rs:41]: No

Comment (zhongzc) [reply src/file-engine/src/manifest.rs:41]: So we can ignore the time index and other indexes for file engine?

Comment (WenyXu) [reply src/file-engine/src/manifest.rs:41]: > So we can ignore the time index and other indexes for file engine?

Yes, We can

File: src/file-engine/src/query.rs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+                        }
+                    } else if let Some(constraint) = region_column_schema.default_constraint() {
+                        constraint
+                            .create_default_vector(region_data_type, true, file_row_count)
+                            .unwrap() // TODO(zhongzc): error
+                    } else {
+                        let mut mutable_vector = region_data_type.create_mutable_vector(1);
+                        mutable_vector.push_null();
+                        let base_vector = mutable_vector.to_vector();
+                        base_vector.replicate(&[file_row_count])
```
Comment (WenyXu) [inline src/file-engine/src/query.rs:179]: Try `NullArray::new`

File: src/file-engine/src/query.rs [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+                let region_schema = self.region_schema.clone();
+                let mut columns = Vec::with_capacity(region_schema.num_columns());
+                for region_column_schema in region_schema.column_schemas() {
+                    let region_data_type = &region_column_schema.data_type;
+
+                    let file_column = file_record_batch.column_by_name(&region_column_schema.name);
+                    let column = if let Some(file_column) = file_column {
+                        if &file_column.data_type() != region_data_type {
+                            file_column.cast(region_data_type).unwrap() // TODO(zhongzc): error
```
Comment (WenyXu) [inline src/file-engine/src/query.rs:167]:
```suggestion
                            file_column.cast(region_data_type).map_err(BoxedError::new).context(common_recordbatch::error::ExteranlSnafu)?;
```

File: src/file-engine/src/query.rs [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+                        } else {
+                            file_column.clone()
+                        }
+                    } else if let Some(constraint) = region_column_schema.default_constraint() {
+                        constraint
+                            .create_default_vector(region_data_type, true, file_row_count)
+                            .unwrap() // TODO(zhongzc): error
```
Comment (WenyXu) [inline src/file-engine/src/query.rs:174]: ditto

**Review by WenyXu — APPROVED** [review]
There are some conflicts

File: src/file-engine/src/engine.rs [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+
+    async fn handle_close(
+        &self,
+        region_id: RegionId,
+        _request: RegionCloseRequest,
+    ) -> EngineResult<Output> {
+        let _lock = self.region_mutex.lock().await;
+
+        let mut regions = self.regions.write().await;
+        if !regions.contains_key(&region_id) {
```
[the remark this answers is not in the document]
Comment (zhongzc) [reply src/file-engine/src/engine.rs:243]: The purpose of writing this is to find a place to put this log: `Try to close region`. Of course if remove this log, the logic here will be simplified. I'll fix it

File: src/file-engine/src/engine.rs [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        }
+    }
+
+    async fn stop(&self) -> EngineResult<()> {
+        let _lock = self.region_mutex.lock().await;
+        self.regions.write().await.clear();
+        Ok(())
+    }
+
+    fn set_writable(&self, _region_id: RegionId, _writable: bool) -> EngineResult<()> {
```
[the remark this answers is not in the document]
Comment (zhongzc) [reply src/file-engine/src/engine.rs:151]: To keep things simple, let's return `Ok(())` directly. Next time find out better semantics about the API.

File: src/file-engine/src/query.rs [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+                            .with_context(|_| CreateDefaultSnafu {
+                                column: scan_column_schema.name.clone(),
+                            })
+                            .map_err(BoxedError::new)
+                            .context(ExternalSnafu)?
+                            .with_context(|| MissingColumnNoDefaultSnafu {
+                                column: scan_column_schema.name.clone(),
+                            })
+                            .map_err(BoxedError::new)
+                            .context(ExternalSnafu)?
```
Comment (WenyXu) [inline src/file-engine/src/query.rs:195]: Can we simplify this?

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**